packages/workflow/src/internals.ts

```diff
@@ -377,35 +384,75 @@ export class State {
    */
   public readonly signalHandlers = new Map<string, WorkflowSignalType>();
 
+  /**
+   * Source map file for looking up the source files in response to __enhanced_stack_trace
+   */
+  public sourceMap: RawSourceMap | undefined;
+
+  /**
+   * Whether or not to send the sources in enhanced stack trace query responses
+   */
+  public showStackTraceSources = false;
+
+  protected getStackTraces(): Stack[] {
+    const { childToParent, promiseToStack } = (globalThis as any).__TEMPORAL__.promiseStackStore as PromiseStackStore;
+    const internalNodes = new Set(
+      [...childToParent.values()].reduce((acc, curr) => {
+        for (const p of curr) {
+          acc.add(p);
+        }
+        return acc;
+      }, new Set())
+    );
+    const stacks = new Map<string, Stack>();
+    for (const child of childToParent.keys()) {
+      if (!internalNodes.has(child)) {
+        const stack = promiseToStack.get(child);
+        if (!stack || !stack.formatted) continue;
+        stacks.set(stack.formatted, stack);
+      }
+    }
+    // Not 100% sure where this comes from, just filter it out
+    stacks.delete('    at Promise.then (<anonymous>)');
+    stacks.delete('    at Promise.then (<anonymous>)\n');
+    return [...stacks].map(([_, stack]) => stack);
+  }
+
   /**
    * Mapping of query name to handler
    */
   public readonly queryHandlers = new Map<string, WorkflowQueryType>([
     [
       '__stack_trace',
       () => {
-        const { childToParent, promiseToStack } = (globalThis as any).__TEMPORAL__
-          .promiseStackStore as PromiseStackStore;
-        const internalNodes = new Set(
-          [...childToParent.values()].reduce((acc, curr) => {
-            for (const p of curr) {
-              acc.add(p);
+        return this.getStackTraces()
+          .map((s) => s.formatted)
+          .join('\n\n');
+      },
+    ],
+    [
+      '__enhanced_stack_trace',
+      (): EnhancedStackTrace => {
+        const { sourceMap } = this;
+        const sdk: SDKInfo = { name: 'typescript', version: pkg.version };
+        const stacks = this.getStackTraces().map(({ structured: locations }) => ({ locations }));
+        const sources: Record<string, FileSlice[]> = {};
+        if (this.showStackTraceSources) {
+          for (const { locations } of stacks) {
+            for (const { filePath } of locations) {
+              if (!filePath) continue;
+              const content = sourceMap?.sourcesContent?.[sourceMap?.sources.indexOf(filePath)];
+              if (!content) continue;
+              sources[filePath] = [
+                {
+                  content,
```

Comment:
What about `\r\n` to `\n` normalization?

Reply:
I wonder if we could somehow make sure that normalization is done in webpack. That would be more efficient but not sure how difficult it might be...

Reply:
TBH I thought about this some more and I don't think it really matters, the UI can handle both line endings.
Didn't bother to update the proposal.